src/firecracker/src/metrics.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::os::unix::io::AsRawFd;
+use std::time::Duration;
+
+use logger::{Metric, LOGGER, METRICS};
+use polly::event_manager::EventHandler;
+use polly::pollable::{Pollable, PollableOp, PollableOpBuilder};
+use timerfd::{ClockId, SetTimeFlags, TimerFd, TimerState};
+
+/// Metrics reporting period.
+pub const WRITE_METRICS_PERIOD_MS: u64 = 60000;
+
+/// Object to drive periodic reporting of metrics.
+pub struct PeriodicMetrics {
+    write_metrics_event_fd: TimerFd,
+    flush_counter: u64,
+}
+
+impl PeriodicMetrics {
+    /// PeriodicMetrics constructor. Can panic on `TimerFd` creation failure.
+    pub fn new() -> Self {
+        let write_metrics_event_fd = TimerFd::new_custom(ClockId::Monotonic, true, true)
+            .expect("Cannot create the metrics timer fd.");
+        PeriodicMetrics {
+            write_metrics_event_fd,
+            flush_counter: 0,
+        }
+    }
+
+    /// Start the periodic metrics engine which will flush metrics every `interval_ms` millisecs.
+    pub fn start(&mut self, interval_ms: u64) {
+        // Arm the log write timer.
+        let timer_state = TimerState::Periodic {
+            current: Duration::from_millis(interval_ms),
+            interval: Duration::from_millis(interval_ms),
+        };
+        self.write_metrics_event_fd
+            .set_state(timer_state, SetTimeFlags::Default);
+
+        // Log the metrics straight away to check the process startup time.
+        self.log_metrics();
+    }
+
+    fn log_metrics(&mut self) {
+        // Please note that, if LOGGER has no output file configured yet, it will write to
+        // stdout, so logging will interfere with console output.
+        if let Err(e) = LOGGER.log_metrics() {
+            METRICS.logger.missed_metrics_count.inc();
+            error!("Failed to log metrics: {}", e);
+        }
+
+        // Only used in tests, but has virtually no cost in production.
+        self.flush_counter += 1;
```

Comment:
nit: #[cfg(test)]

Reply:
You can wrap it in a code block instead:
```
#[cfg(test)] {
            self.flush_counter += 1;
        }
```